### PR TITLE
Smelter improvements

### PR DESCRIPTION
**Pull request by covertcorvid**
## About The Pull Request
Ports part of https://github.com/BeeStation/NSV13/pull/1643
Adds ID variable linking for smelting and stacking machines and their consoles, and lets miners redeem points from the smelter.

## Why It's Good For The Game
The ID links let mappers auto-link the machines to their consoles even if they're not next to each other.
Point redemption makes the smelter actually somewhat usable as an ORM alternative.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

smelter

</details>

## Changelog
:cl: Karmic-Skink
add: The smelter console now allows you to obtain mining points
code: Mappers can use the link_id variables on the smelting and stacking machines and consoles to automatically link machines that are not next to each other
/:cl:

### CODE REVIEW

File: code/modules/mining/machine_processing.dm [hunk 1]
```diff
@@ -57,14 +57,25 @@
...
 	else
-		return INITIALIZE_HINT_QDEL
+		machine = locate(/obj/machinery/mineral/processing_unit, get_step(src, machinedir))
+		if (machine)
+			machine.CONSOLE = src
+		else
+			return INITIALIZE_HINT_QDEL
+
+/obj/machinery/mineral/processing_unit_console/LateInitialize()
+	if(link_id) //If mappers set an ID)
```
Comment (Archanial) [inline code/modules/mining/machine_processing.dm]:
```suggestion
```
there is no possiblity that we end up here with no link id

File: code/modules/mining/machine_processing.dm [hunk 2]
```diff
@@ -57,14 +57,25 @@
...
+		for(var/obj/machinery/mineral/processing_unit/PU in GLOB.machines)
+			if(PU.link_id == link_id)
+				machine = PU
+				machine.CONSOLE = src
```
Comment (Archanial) [inline code/modules/mining/machine_processing.dm]: early return this loop

File: code/modules/mining/machine_processing.dm [hunk 3]
```diff
@@ -97,6 +108,17 @@
...
+	if(href_list["redeem"])
+		var/mob/M = usr
+		var/obj/item/card/id/I = M.get_idcard(TRUE)
+		if(machine.points)
+			if(I?.mining_points += machine.points)
+				machine.points = 0
+			else
+				to_chat(usr, "<span class='warning'>No ID detected.</span>")
+		else
+			to_chat(usr, "<span class='warning'>No points to claim.</span>")
```
Comment (Archanial) [inline code/modules/mining/machine_processing.dm]:
```suggestion
		if(!I)
			to_chat(usr, "<span class='warning'>No ID detected.</span>")
			return
		if(!machine.points)
			to_chat(usr, "<span class='warning'>No points to claim.</span>")
		I.mining_points += machine.points
		machine.points = 0
```

File: code/modules/mining/machine_processing.dm [hunk 4]
```diff
@@ -140,13 +164,28 @@
...
 
 /obj/machinery/mineral/processing_unit/proc/get_machine_data()
 	var/dat = "<b>Smelter control console</b><br><br>"
+
+	//On or off
+	dat += "Machine is currently "
+	if (on)
+		dat += "<A href='?src=[REF(CONSOLE)];set_on=off'>On</A> "
+	else
+		dat += "<A href='?src=[REF(CONSOLE)];set_on=on'>Off</A> "
```
Comment (Archanial) [inline code/modules/mining/machine_processing.dm]:
```suggestion
	dat += "Machine is currently [ on ? "<A href='?src=[REF(CONSOLE)];set_on=off'>On</A> " : "<A href='?src=[REF(CONSOLE)];set_on=on'>Off</A> "]
```

Comment (Archanial) [reply code/modules/mining/machine_processing.dm]: Actually why isn't a toggle? The function itself should check for on var and act on it, not aact on some href.

File: code/modules/mining/machine_stacking.dm [hunk 5]
```diff
@@ -9,12 +9,23 @@
...
+		machine = locate(/obj/machinery/mineral/stacking_machine, get_step(src, machinedir))
+		if (machine)
+			machine.console = src
+
+/obj/machinery/mineral/stacking_unit_console/LateInitialize()
+	if(link_id) //If mappers set an ID)
+		for(var/obj/machinery/mineral/stacking_machine/SM in GLOB.machines)
+			if(SM.link_id == link_id)
+				machine = SM
+				machine.console = src
```
Comment (Archanial) [inline code/modules/mining/machine_stacking.dm]:
```suggestion
	if(!link_id) //If mappers set an ID)
		return
	for(var/obj/machinery/mineral/stacking_machine/SM in GLOB.machines)
		if(SM.link_id == link_id)
			machine = SM
			machine.console = src
			return
```

**Review by Archanial — DISMISSED** [review]

File: code/modules/mining/machine_processing.dm (hunk 4 above)
Comment (covertcorvid) [reply code/modules/mining/machine_processing.dm]: As in move the on/off change onto the machine instead of the console?

**Comment by PowerfulBacon** [comment]
Won't this let prisoners at the gulag get out twice as fast?

**Comment by covertcorvid** [comment]
> Won't this let prisoners at the gulag get out twice as fast?

Genuinely forgot about the special gulag stacker. Would it be better to remove points from the stacker redemption console, or have a special labor camp smelter that doesn't have points?

File: code/modules/mining/machine_processing.dm (hunk 4 above)
Comment (Archanial) [reply code/modules/mining/machine_processing.dm]: Yes, the function itself should be a toggle, that checks if the machine is on or off

**Comment by PowerfulBacon** [comment]
I don't mind

**Comment by covertcorvid** [comment]
Is it intentional that, when using the labor camp stacking machine, if you're able to get access to the cockpit somehow you can just go grab the mats and put them through again for more points? If that's desired or useful behavior I'll just make a labor camp subtype for the smelting and keep the system as it is.

Otherwise, I can either update the machines to separately handle gulag and mining points, or make the whole system run on mining points. Since prisoners get given prisoner IDs, the only obvious drawback to using mining points would be the possibility of miners somehow transferring points, but tbh if a miner wanted to break out a prisoner they already could.

**Comment by PowerfulBacon** [comment]
I think its intentional, I just kind of assumed the ores go straight to the silo. Any solution is fine as long as from the prisoner side of things its basically the same (put minerals in, get points, leave) and it takes the same amount of time as before (point rate isn't doubled and you can't do an exploit to get infinite points or something by prisoners removing iron from the machine.)

**Comment by covertcorvid** [comment]
Added a laborcamp smelter to keep the system as is so they don't get the points and then build a bunch of statues or something equally pointless

**Review by PowerfulBacon — APPROVED** [review]

**Pull request merged**